src/main/java/graphql/nadel/engine/HydrationInputResolver.java

```diff
@@ -489,13 +494,25 @@ private Field createBatchHydrationTopLevelField(ExecutionContext executionContex
         ListExecutionResultNode listResultNode = (ListExecutionResultNode) root;
         List<ExecutionResultNode> resolvedNodes = listResultNode.getChildren();
 
+        if (isResolveByIndex) {
+            assertTrue(resolvedNodes.size() == hydrationInputNodes.size(), () -> "Expect the size of the resolved nodes to match the size of the input when resolving by index");
+        }
+
         List<ExecutionResultNode> result = new ArrayList<>();
         Map<String, FieldTransformation> transformationByResultField = queryTransformationResult.getFieldIdToTransformation();
         Map<String, String> typeRenameMappings = queryTransformationResult.getTypeRenameMappings();
 
         boolean first = true;
-        for (HydrationInputNode hydrationInputNode : hydrationInputNodes) {
-            ObjectExecutionResultNode matchingResolvedNode = findMatchingResolvedNode(executionContext, hydrationInputNode, resolvedNodes);
+        for (int i = 0, hydrationInputNodesSize = hydrationInputNodes.size(); i < hydrationInputNodesSize; i++) {
+            HydrationInputNode hydrationInputNode = hydrationInputNodes.get(i);
+
+            ExecutionResultNode matchingResolvedNode;
+            if (isResolveByIndex) {
+            matchingResolvedNode = resolvedNodes.get(i);
```

Comment:
we should fail clearly hear with a clear exception telling what failed: e.g. underlying service doesn't honor the contract etc.

Reply:
Under which cases? I do check above that the size of the resolved nodes + the size of the input are identical.

Reply:
I missed that ... that is exactly what I wanted.